### ISSUE TITLE
Unions of $PropertyType with generics not working as expected

### ISSUE THREAD

**Issue by nrabinowitz** [facebook/flow#7436]
Flow version: 0.92

Using `$PropertyType` with generics in a union can cause errors in which only the first element of the union is being matched, particularly when typing the output of a function.

The following code, without generics, works correctly with no errors ([Try Flow](https://flow.org/try/#0MYGwhgzhAEAqCmEAusCeAHR0DeAoa0AdgK4C2ARvAE4Bc0A5CRdQIz0Dc+0yVAloQHM69HvwFtOAX1y4kGeNABiveCAAm0ALxcAPtAAkABSoB7TFTlpMAHgTIriADQNRg+gD5dB42eqX5togo8hDOjGSUVB6cuABmxITASLwmhNyq8EkAFFzwAB7oVAAyvMh0yqpquACU5SrqOFxU8EjEVGn5hSXIUjJZIkh8gmx1ldWc-UyRI0r1auO4WRAZ2QND4vS1s2MTyyCZSJMRrJuj6gsyuEA)):
```
class TestTypes {
  number: 'number1';
  string: 'string1';
}

type Field =
  | $PropertyType<TestTypes, 'string'>
  | $PropertyType<TestTypes, 'number'>;

function select(
  exprList: Field
): Field {
  return exprList;
}

('string1': Field);
('number1': Field);
(select('string1'): Field);
(select('number1'): Field);
```

When generics are used instead, the `select` function cannot be typed correctly for any member of the union *except the first* ([Try Flow](https://flow.org/try/#0C4TwDgpgBAYglhANgEwDwGUAqA+KBeAKCigB8oASABQCcB7Sa0TcCDTAGigHIBnYauADsA5l2xFSFGvQiMQzSG05dBAVwC2AI1liA3AQIAzVYIDGwOLUFQeSCObbYAFBIgAPMNQAycPgC5YBBRHAgBKAPgkNCxcAG8JaghgVWprd08fPn0AXwNTRABDHh4oTAg+BXKoeOI1LVkAlQ1tagBGLn1iPgERRu6hYXacgwB6EagAVR4BqGAAC2hIlCgTS2tkOETzRBAoAHdaagBrEsMhCAA6AidefgH2iKC0MoqWHmxQ-Ru6lofAqNQL2AlXen1G4ymM3miyeK0Ea32cHmswWUDO1D4UHUEHq1CgYCKtmQBDGs1oUAKaJM5gRB2Op3OVyctkQ9mAN36Ina4X+wSBII++hJEOmIhRMKicNpSLmFMEu2xuKgtGhePmBWs0OFaM2mIJxQgyDJFKpZgsVjRBTgiB4TJZbO+zVk3MeAP5b0FBgMQA)):
```
type Field<ST> =
  | $PropertyType<ST, 'string'>
  | $PropertyType<ST, 'number'>;

function select<ST>(
  exprList: Field<ST>
): Field<ST> {
  return exprList;
}

class TestTypes {
  number: 'number1';
  string: 'string1';
}

// Using the Field union directly works fine.
('string1': Field<TestTypes>);
('number1': Field<TestTypes>);

// Using the Field union with the first member passed
// to a function works fine.
(select('string1'): Field<TestTypes>);

// Using the Field union with any member other than the
// first passed to a function fails.
(select('number1'): Field<TestTypes>);
// ^ Cannot cast `select(...)` to `Field` because string [1] is incompatible with string literal `string1` [2] in type argument `ST` [3].
```

The fact that this works without error for the first member of the union and not for subsequent members suggests this is almost certainly a bug.

**Comment by jps327** [comment]
Your first example is unsound, which is why when you try and extend it to use generics things don't work.
If in your first example if we were to do:
```
const test1: 'string1' = select('string1');
```
You'd think this should work, but it does not. You get a type error. Because Flow does not recognize that if you passed in 'string1' that the output should also be 'string1'.
[Example](https://flow.org/try/#0MYGwhgzhAEAqCmEAusCeAHR0DeAoa0AdgK4C2ARvAE4Bc0A5CRdQIz0Dc+0yVAloQHM69HvwFtOAX1y4kGeNABiveCAAm0ALxcAPtAAkABSoB7TFTlpMAHgTIriADQNRg+gD5dB42eqX5togo8hDOjGSUVB6cuABmxITASLwmhNyq8EkAFFzwAB7oVAAyvMh0yqpquACU5SrqOFxU8EjEVGn5hSXIUjJZIkh8gmx1ldWc-UyRI0r1auO4WRAZ2QND4vS1s2MTyyCZSJMRrJuj6gu4wKnI0EhBLMKuG1rp+6tPbAtAA)

The reason is because your `select` function is missing any generics. This is correct:
```
function select<F: Field>(
  exprList: F
): F {
  return exprList;
}
```
[And now you see that it works.](https://flow.org/try/#0MYGwhgzhAEAqCmEAusCeAHR0DeAoa0AdgK4C2ARvAE4Bc0A5CRdQIz0Dc+0yVAloQHM69HvwFtOAX1y4kGeNABiveCAAm0ALxcAPtAAkABSoB7TFTlpMAHgTIriADQNRg+gD5dB42eqX5togo8hDOjGSUVB6cuABmxITASLwmhNyq8EnWinTKqmruABRc8AAe6FQAMrzIubgAlLk4XFTwSMRUaWUV1chSMoUiSHyCbLkq6vWcg0yRY0oTalO4hRAZSYOu4vSNC-nLq+tIMxGsO+P7McCpyNBIQSzCW2xa6SCZx0Mj28tAA)

What about your example with generics then?
This works:
```
function select<ST: { +[string]: any }, V: $Values<ST>>(
  exprList: V
): V {
  return exprList;
}
```
Basically, you need to tell it that `select` should expect a value `V` which can only be one of the `$Values` from `ST`. [This works now](https://flow.org/try/#0C4TwDgpgBAYglhANgEwDwGUAqA+KBeAKCigB8oASABQCcB7Sa0TcCDTAGigHIBnYauADsA5l2xFSFGvQiMQzSG05dBAVwC2AI1liA3AQIAzVYIDGwOLUFQeSCObYAuKAG8oAagDafASIC6zgCGgiBQAL6cAGrO5JGBiKoQPGzY2AAUEhAAHmDUADJwfM6RBACUxa4S1BDAqtTW2bkFfPphBqaIgTw8UJhJwApJlcRqWrLOKhra1ACMXPrEPkLCE0sic60GAPRbUACqPMtQwAAW0PBIyFAmltbIcNXmiKEA7rTUANY9hkIQAHQENK8fjLObOC4oVB9PiDHjYUr6IGjaZg2AISHQgYsOEI7a7A5HU7ndFXG5WKAvOCnY5nKA-ah8KDqCBjahQMBdWzIAg7Y60KCBOkmcy3CnvL5034AtK2RD2YBAtbCOblNGXKH9WHw-Q8-GHEQ04mXa6CUWU6nBULM1lQWhEtmnYKG3WShnAdmciBXYD8wXGMwWcmGQJwRA8aWy+VIqayFXgkkamHY7XtKyM4D9GarEHrLj4Gx2cyKnPKrgIoA).

Hope this helps!

**Comment by SamChou19815** [comment]
They now work under `inference_mode=lti`
